batchproducer/batchproducer.go

```diff
@@ -187,15 +200,21 @@ func (b *batchProducer) Add(data []byte, partitionKey string) error {
 
 // from/for interface Producer
 func (b *batchProducer) Start() error {
-	if b.isRunning() {
-		return nil
+	b.runningMu.Lock()
+	defer b.runningMu.Unlock()
```

Comment:
This is throwing full exclusion contention around the entire method. We can reduce this to just:

```go
b.runningMu.Lock()
if b.running {
  b.runningMu.Unlock()
  return ErrAlreadyStarted
}
b.running = true
b.runningMu.Unlock()
```

This won't make much difference in the end as spinning off a go routine and waiting for it to start won't take that long (for now, anyway.) but it feels like unnecessary contention.